packages/no-trapping-shim/src/no-trapping-pony.js

```diff
@@ -0,0 +1,206 @@
+ * `Reflect.suppressTrapping`.
+ *
+ * @param {any} specimen
+ * @param {boolean} shouldThrow
+ * @returns {boolean}
+ */
+const suppressTrappingInternal = (specimen, shouldThrow) => {
+  if (noTrappingSet.has(specimen)) {
+    return true;
+  }
+  freeze(specimen);
+  if (!proxyHandlerMap.has(specimen)) {
+    noTrappingSet.add(specimen);
+    return true;
+  }
+  const [target, handler] = proxyHandlerMap.get(specimen);
+  if (isNoTrappingInternal(target, shouldThrow)) {
+    noTrappingSet.add(specimen);
+    return true;
+  }
+  const trap = handler.suppressTrapping;
+  if (trap === undefined) {
+    const result = suppressTrappingInternal(target, shouldThrow);
+    if (result) {
+      noTrappingSet.add(specimen);
+    }
+    return result;
+  }
+  const result = apply(trap, handler, [target]);
+  const ofTarget = isNoTrappingInternal(target, shouldThrow);
+  if (result !== ofTarget) {
+    if (shouldThrow) {
+      throw TypeError(
+        `'suppressTrapping' proxy trap does not reflect 'isNoTrapping' of proxy target (which is '${ofTarget}')`,
+      );
+    }
+    return false;
+  }
+  if (result) {
+    noTrappingSet.add(specimen);
+  }
+  return result;
+};
+
+export const extraReflectMethods = freeze({
+  isNoTrapping(target) {
+    if (isPrimitive(target)) {
+      throw TypeError('Reflect.isNoTrapping called on non-object');
+    }
+    return isNoTrappingInternal(target, false);
+  },
+  suppressTrapping(target) {
+    if (isPrimitive(target)) {
+      throw TypeError('Reflect.suppressTrapping called on non-object');
+    }
+    return suppressTrappingInternal(target, false);
+  },
+});
+
+export const extraObjectMethods = freeze({
+  isNoTrapping(target) {
+    if (isPrimitive(target)) {
+      return true;
+    }
+    return isNoTrappingInternal(target, true);
+  },
+  suppressTrapping(target) {
+    if (isPrimitive(target)) {
+      return target;
+    }
+    if (suppressTrappingInternal(target, true)) {
+      return target;
+    }
+    throw TypeError('preventExtensions trap returned falsy');
+  },
+});
+
+const addExtras = (base, ...extrasArgs) => {
+  for (const extras of extrasArgs) {
+    for (const key of ownKeys(extras)) {
+      if (base[key] !== extras[key]) {
+        defineProperty(base, key, {
+          value: extras[key],
+          writable: true,
+          enumerable: false,
+          configurable: true,
+        });
+      }
+    }
+  }
+};
+
+const ReflectPlus = {};
+addExtras(ReflectPlus, OriginalReflect, extraReflectMethods);
+export { ReflectPlus };
+
+const ObjectPlus = function Object(...args) {
+  if (new.target) {
+    return construct(OriginalObject, args, new.target);
+  } else {
+    return apply(OriginalObject, this, args);
+  }
+};
+ObjectPlus.prototype = OriginalObject.prototype;
+addExtras(ObjectPlus, OriginalObject, extraObjectMethods);
+export { ObjectPlus };
+
+const makeMetaHandler = handler =>
+  freeze({
+    get(_, trapName, _receiver) {
+      return freeze((target, ...rest) => {
```

Comment:
The overhead of creating a new closure for every trap invocation feels scary. If the goal is to catch future traps not yet specified, why not use a handler with well known traps defined in advance and a proxy as proto to dynamically define unknown traps?

Reply:
Given the second part of your suggestion "a proxy as proto to dynamically define unknown traps?", I don't need to do the first part "why not use a handler with well known traps defined in advance". Rather, I have the "proxy as proto" (aka the metaHandler) install overrides on the handlerPlus lazily, avoiding further uses of the double reflection.

So, the second part of your suggestion, done. ***PLEASE*** rereview this carefully, as it is much trickier than the previous non-lazy version.

Resolving.